backend/remote-state/http/backend_state.go

```diff
@@ -0,0 +1,177 @@
+		}
+	}
+
+	sort.Strings(result[1:])
+	if sort.SearchStrings(result, backend.DefaultStateName) == 0 {
+		result = append(result, backend.DefaultStateName)
+	}
+	return result, nil
+}
+
+// DeleteState deletes a state file
+func (b *Backend) DeleteState(name string) error {
+	if name == backend.DefaultStateName || name == "" {
+		return fmt.Errorf("can't delete default state")
+	}
+	client, err := b.remoteClient(name)
+	if err != nil {
+		return err
+	}
+
+	return client.Delete()
+}
+
+// get a remote client configured for this state
+func (b *Backend) remoteClient(name string) (*RemoteClient, error) {
+	if name == "" {
+		return nil, errors.New("missing state name")
+	}
+	client := &RemoteClient{
+		client:        b.client,
+		address:       b.statePath(name),
+		updateMethod:  b.updateMethod,
+		lockAddress:   b.lockPath(name),
+		unlockAddress: b.lockPath(name),
+		lockMethod:    b.lockMethod,
+		unlockMethod:  b.unlockMethod,
+		username:      b.username,
+		password:      b.password,
+	}
+	return client, nil
+}
+
+// State reads the state file
+func (b *Backend) State(name string) (state.State, error) {
+	client, err := b.remoteClient(name)
+	if err != nil {
+		return nil, err
+	}
+
+	stateMgr := &remote.State{Client: client}
+
+	// Grab the value
+	if err := stateMgr.RefreshState(); err != nil {
+		return nil, err
+	}
+
+	// If we have no state, we have to create an empty state
+	if v := stateMgr.State(); v == nil {
+		// take a lock on this state while we write it
+		lockInfo := state.NewLockInfo()
+		lockInfo.Operation = "init"
+		lockID, err := client.Lock(lockInfo)
+
+		if err != nil {
+			return nil, fmt.Errorf("failed to lock http state: %s", err)
+		}
+
+		// Local helper function so we can call it multiple places
+		lockUnlock := func(parent error) error {
+			if err := stateMgr.Unlock(lockID); err != nil {
+				return fmt.Errorf(strings.TrimSpace(errStateUnlock), lockID, err)
+			}
+			return parent
+		}
+
+		if err := stateMgr.WriteState(terraform.NewState()); err != nil {
+			err = lockUnlock(err)
+			return nil, err
+		}
+		if err := stateMgr.PersistState(); err != nil {
+			err = lockUnlock(err)
+			return nil, err
+		}
+
+		// Unlock, the state should now be initialized
+		if err := lockUnlock(nil); err != nil {
+			return nil, err
+		}
+	}
+	return stateMgr, nil
+}
+
+// Construct the path of the state file based on named state
+func (b *Backend) statePath(name string) string {
+	paths := []string{b.address, "/", name, stateFileSuffix}
+	var buf bytes.Buffer
+	for _, p := range paths {
+		buf.WriteString(p)
+	}
+	path := buf.String()
+
+	return path
+}
+
+// Construct the path of the lock file based on named state
+func (b *Backend) lockPath(name string) string {
+	paths := []string{b.address, "/", name, lockFileSuffix}
+	var buf bytes.Buffer
+	for _, p := range paths {
+		buf.WriteString(p)
+	}
+	path := buf.String()
+
+	return path
+}
+
+// Client initialized
+func (b *Backend) Client() *RemoteClient {
```

Comment:
Why do you need both the `Client()` and `remoteClient()` functions? This one doesn't even really do anything... If anything, shouldn't this call `remoteClient()` so the client is actually initialized. The comment is a lie.

Reply:
that was a leftover. thanks for catching it.